### ISSUE TITLE
Missing ability to include another YAML file in a short way (ie combine load + func call)

### ISSUE THREAD

**Issue by mildred** [vmware-tanzu/carvel-ytt#18]
In ytt, there is the possibility to load values from files, but it does not seem that it is possible to load another YAML file from the main ytt file.

Could this be considered as an addition?

Something like this:

main.yaml
```yaml
---
test: #@ include("hello.yaml")
```

hello.yaml
```yaml
---
#@ def world()
"world"
#@ end

hello: #@ world()
```

the result would be:

```yaml
---
test:
  hello: "world"
```

Having a way to pass variables to the included file would be nice. What this is is basically a `#@ load()` but with the function definition implicit in the included file.

**Comment by nimakaviani** [comment]
hi @mildred ,

ytt in fact allows for loading files with function definitions in other files.

Here is an example from the ytt playground with a very similar usecase to what you explained above: https://get-ytt.io/#example:example-load

**Comment by mildred** [comment]
Yes, I understand, but I wanted a way to avoid having to explicitly declare the functions. I would like to have instead of `funcs.lib.yml`

**`func.lib.yml`:**
```yaml
name: max
cities:
- SF
- LA
```

*without* the `#@ def func1():`. The whole file would be by itself a single function.

Perhaps that can be made into the `load()` builtin. So when there is no `#@ def` in the loaded file, an implicit `_main` function could be defined.

The main template could then be:

**`main.yml`**
```yaml
#@ load("func.lib.yml", func1="_main")

func1_key: #@ func1()
```

which would result in:

```yaml
func1_key:
  name: max
  cities:
  - SF
  - LA
```

It's just that I'm used to have include statements, and having to define functions seems a bit odd when you mean to have the whole file as a single function.

**Comment by cppforlife** [comment]
i was actually considering adding this functionality at some point, but went with a more conservative approach for now.

this was due to awkwardness with parametarization. ie how does one specify which args whole file takes. i guess it might be worth adding this feature without allowing args, but then "upgrade" from no-args to args is a bit weird. wdyt?

**Comment by mildred** [comment]
You're right, what I'm asking is not really necessary, and is more convenience. in any case it's not very important.

Upgrade can happen simply by adding a `#@ def` statement enclosing the whole file that map to the default function name.

**Comment by mildred** [comment]
feel free to close this issue or file it for later if you think it's relevant.

**Comment by mildred** [comment]
What can be useful though, is having a way to perform both the load() and the function call.

A way to shorten the following file:

```yaml
#@ load("parts/a.yml", parta="main")
#@ load("parts/a.yml", partb="main")
#@ load("parts/a.yml", partc="main")
---
#@ parta()
---
#@ partb()
---
#@ partc()
```

**Comment by mildred** [comment]
@cppforlife, how do you suggest this should be implemented ?

I noticed that `load()` with only the file name specified was currently invalid (*load statement must import at least 1 symbol*). I suggest to change `load()` from a statement to an expression (it would return a value) the following way:

- `load()` with no imports defined is returning a function corresponding to the complete loaded document
- `load()` with a more imports, in addition to importing the functions in the current scope, would return a dict containing the function objects indexed by their function name

I suggest `load()` with no arguments should be equivalent to enclosing the loaded file with:

```yaml
#@def the_anonymous_main_function(*args, **kwargs):
...
#@end
```

(and returning the function)

**Comment by mildred** [comment]
Oh, I didn't see `load()` was part of Starlark, and as such it might not be possible to convert it to an expression... too bad.

**Comment by cppforlife** [comment]
yeah `load()` is part of starlark so it might have to be something else if we continue on this route.

i think this issue actually relates to a different problem ive been thinking about recently. given that ytt supports custom libraries (examples: https://get-ytt.io/#example:example-load-custom-library, https://github.com/k14s/k8s-lib), there should be facilities to reuse non-library files (ie templates) from a library.

in the example linked above, k8s-lib, offers library with `make()` method that returns another struct. this is fine as that repo was designed to be used as a library. however, in other cases, repo may not be designed to be used as a library and just have bunch of templates (no function to call to get them). currently if those templates are "imported" via `_ytt_lib` there would be no way to access them or pass in specific data/values to them.

i was imagining adding something like below, which may solve this issue as well... this example tries to use https://github.com/k14s/k8s-simple-app-example/tree/develop/config-step-2-template as a library:

```yaml
#@ load("@ytt:lib", "lib")
#@ load("@ytt:template", "template")

#! no custom data/values
#@ sample_app1 = lib.load("github.com/k14s/k8s-simple-app-example/config-step-2-template")

#! with custom data/values
#@ sample_app2 = lib.load("github.com/k14s/k8s-simple-app-example/config-step-2-template", {"hello_msg": "another-msg"})

--- #@ template.replace(sample_app1.templates())
--- #@ template.replace(sample_app2.templates())
```

and for this issue:

```yaml
#@ load("@ytt:lib", "lib")
stuff: #@ lib.self().template("hello.yaml")
```

one challenge with `stuff: #@ lib.self().template("hello.yaml")` though, is that templating hello.yaml returns document set, which cannot be value of a map item. one way to solve that would be to add `unwrap()` (or something similar) to yamlfrag so it will end up like this: 

```yaml
stuff: #@ lib.self().template("hello.yaml").unwrap()
#! or
stuff: #@ lib.self().unwrapped_template("hello.yaml")
#! or 
stuff: #@ lib.self().inline("hello.yaml")
```

**Comment by mildred** [comment]
That sounds like a good idea if you believe it's possible to load files outside of Starlark load statement.

The `unwrap()` might not be necessary if yaml fragments are indexable. Just `.template("hello.yml")[0]` might suffice.

**Comment by mildred** [comment]
A way to pass values to loaded templates, would be to use the existing `@ytt:data` library. inside the loaded template, `data.values.FOO` can resolve to:

- `FOO` defined when loading the template via `lib.load("lib/", {"FOO": "bar"}).template("hello.yml")`
- `FOO` defined in a `#@data/values` annotated file in the `lib/` library directory

reusing existing mechanisms and providing composability.

**Comment by cppforlife** [comment]
> reusing existing mechanisms and providing composability

yup, that's what i was thinking as well. though need to also figure out how end user can customize data/values given to that library. top level template's data/values can be customized via --data-value flags. i've noted down some thinking in here about using overlays as well: https://github.com/k14s/ytt/issues/11.

one challenge remains is how do you reference particular data/values since library can be templated multiple times (in example above, we have sample_app1 and sample_app2). nothing good came to mind so far...

**Comment by mildred** [comment]
Do you mind if I attempt an implementation of that ? (implement @ytt:lib.load, extract code from `pkg/cmd/template/cmd.go` in a separate package to reuse it for load(0 implementation)

**Comment by cppforlife** [comment]
@mildred sure, go ahead!

**Comment by mildred** [comment]
See WIP: https://github.com/k14s/ytt/compare/develop...mildred:ytt-lib

**Comment by mildred** [comment]
![image](https://user-images.githubusercontent.com/33804/58745316-bb3b0180-844f-11e9-9675-1815fc8c8523.png)

**Comment by mildred** [comment]
![image](https://user-images.githubusercontent.com/33804/58745482-ddce1a00-8451-11e9-8f1d-4f446faf8e60.png)

**Comment by cppforlife** [comment]
@mildred do you want to start a PR. i'd like to leave a few comments as i look deeper into the code.

**Comment by cppforlife** [comment]
oops. didnt notice you've already done that...

**Comment by mildred** [comment]
overriding values:

![image](https://user-images.githubusercontent.com/33804/58749490-556a6c00-8487-11e9-91b2-3a5976514dfe.png)

**Comment by cppforlife** [comment]
after spending morning with @nimakaviani we have came to the following proposal:

```
module == single file (can export variables, functions, or be templated => some type of result e.g. yaml structure, or string, or None)
package == single directory, contains modules
library == collection of packages
```

load path format:

```
[@[library]:][package/]module

with libraries being:
- "" (empty) for current library
- "std" (or "ytt) for standard library provided
- ".+" for any library path under _ytt_lib directory

any number of packages in path
single module at the end (std library modules dont have an extension)
```

examples covering all variations:

```yaml
# find module within current directory (ie relative path, does not support '..')
funcs.yml

# find module within package 'dir' that's inside current package
dir/funcs.yml

# same as @ytt:overlay
@std:overlay

# use "current library" and load particular module
@:common/stuff.yml

# use "github.com/k14s/k8s-lib/app" library and load 
# particular module from it; this library must be 
# located under _ytt_lib (eg _ytt_lib/github.com/k14s/k8s-lib)
@github.com/k14s/k8s-lib/app:module.yml
```

note that `@:common/stuff.yml` case covers the ability to include modules from packages relative to current library root (instead of supporting symlinks or `../..` in paths). for example `load("@:common/funcs.lib.yml", "common1")` would work from `app1/config.yml`, assuming that directory structure is:

```
app1/config.yml
common/funcs.lib.yml
```

this loading behaviour can also be used with `library` module from `std` library (included via `load("@std:library", "library")`).

```python
# loading library object into a variable
a = library.get("")
a = library.get("std") # not supported
a = library.get("github.com/k14s/k8s-lib/app")

# library object has following methods
.data_values({...}) # returns new library object with added data values overrides (takes yaml structure as well for input)
.yaml(["common/stuff.yml"]) # returns yaml structures from "common/stuff.yml"
.text(["common/stuff.yml"]) # returns templates string from "common/stuff.yml"
.funcs("common/stuff.yml") # returns module that has exported methods (see module package)
.list() # ["file1.yml", "file2.yml"]
.read("file1.yml") # <raw contents of file1.yml>
```

what do you think @mildred ?

**Comment by mildred** [comment]
This looks like a good proposal, I have a few questions:

If I'm located in a module in `/_ytt_lib/lib1/pkg1/mod1.yaml` (`@lib1:pkg1/mod1.yaml`)

- Would `@:pkg2/mod2.yaml` reference `/_ytt_lib/lib1/pkg2/mod2.yaml` or `/pkg2/mod2.yaml`?
- Would `@lib2:pkg1/mod1.yaml` reference `/_ytt_lib/lib2/pkg1/mod1.yaml` or `/_ytt_lib/lib1/_ytt_lib/lib2/pkg1/mod1.yaml` or `/_ytt_lib/lib1/pkg1/_ytt_lib/lib2/pkg1/mod1.yaml`, or perhaps any of these in a specific search order?

Also, in your API proposal:

```
.yaml(["common/stuff.yml"]) # returns yaml structures from "common/stuff.yml"
.text(["common/stuff.yml"]) # returns templates string from "common/stuff.yml"
.funcs("common/stuff.yml") # returns module that has exported methods (see module package)
```

Why does `yaml` takes an array and `func` does not? Why not take varargs in the function (less syntax cluttering, but not important)? Also, a feature I wanted in this module was the ability to get all the text or all the yaml from all files from the library if no argument was provided. it frees the library user from knowing the library internals.

**Comment by mildred** [comment]
In particular, if already within `/_ytt_lib/lib1/pkg1/mod1.yaml` (`@lib1:pkg1/mod1.yaml`), if `@lib2:pkg1/mod1.yaml` can reference `/_ytt_lib/lib2/pkg1/mod1.yaml` (either directly or through fallback) it allows more powerful things. In particular it allows to avoid duplication of libraries used in different places

**Comment by nimakaviani** [comment]
I believe

- `@:pkg2/mod2.yaml` reference `/_ytt_lib/lib1/pkg2/mod2.yaml` as the context for `@:` is the _current library_, which in this case is `lib1`.
- same with `@lib2:pkg1/mod1.yaml` referencing  `/_ytt_lib/lib1/pkg1/_ytt_lib/lib2/pkg1/mod1.yaml`

yes we are going to have duplication. However, from the slack discussion we had [here](https://kubernetes.slack.com/archives/CH8KCCKA5/p1562103361366000?thread_ts=1561704953.311700&cid=CH8KCCKA5), I believe we kinda settle on the fact that in the absence of mounting, maybe we can continue with duplication? so having duplicated - but potentially different version of a given library - in multiple `_ytt_lib` directories is probably ok.

> Why does yaml takes an array and func does not?

it is probably an oversight. or did you have particular thoughts @cppforlife ?

> Why not take varargs in the function (less syntax cluttering, but not important)

+1 on using variadics here. makes it cleaner.

> Also, a feature I wanted in this module was the ability to get all the text or all the yaml from all files from the library if no argument was provided. it frees the library user from knowing the library internals.

I prefer things being explicit. makes readability a lot easier. Recently, I made some concourse adjustments using ytt and the team that used the templates were super happy about having explicit references as to where to look for what function or yaml structure. I can see using something like `.yaml(["common/*.yml"])` being a convenience when writing templates but reading those is going to make life harder for some people. Again this is from first hand feedback I received from some other engineers on a project I am contributing to.

**Comment by cppforlife** [comment]
> If I'm located in a module in `/_ytt_lib/lib1/pkg1/mod1.yaml` (`@lib1:pkg1/mod1.yaml`)
> - Would `@:pkg2/mod2.yaml` reference `/_ytt_lib/lib1/pkg2/mod2.yaml` or `/pkg2/mod2.yaml`?
> - Would `@lib2:pkg1/mod1.yaml` reference `/_ytt_lib/lib2/pkg1/mod1.yaml` or `/_ytt_lib/lib1/_ytt_lib/lib2/pkg1/mod1.yaml` or `/_ytt_lib/lib1/pkg1/_ytt_lib/lib2/pkg1/mod1.yaml`, or perhaps any of these in a specific search order?

`@:` means root of current library. and definition of current library depends on how it was loaded from whoever loaded it. with following files:

```
_ytt_lib/github.com/lib1/pkg1/mod1.yaml
_ytt_lib/github.com/lib1/pkg2/mod2.yaml
```

`@:lib1/pkg2/mod2.yaml` will reference `mod2.yaml` if `mod1.yml` was loaded via`@github.com/lib1:pkg1/mod1.yml`.

if `@github.com:lib1/mod1.yaml` was used then current library (`@:`) would be rooted at `_ytt_lib/github.com` (instead of `github.com/lib1`) which would cause `@:pkg2/mod2.yml` not to resolve (from inside `mod1.yaml`).

the reason why we need to depend on the "initiating" load is because we do not have any filesystem markers that signify where one library begins vs ends (`github.com/some-name` is up to interpretation to mean `github.com` library with `some-name` as a package vs `github.com/some-name` is the library.)

> ... or perhaps any of these in a specific search order

all load locations always point to one thing, ie not searching thru multiple directories etc.

Also, in your API proposal:

```
.yaml(["common/stuff.yml"]) # returns yaml structures from "common/stuff.yml"
.text(["common/stuff.yml"]) # returns templates string from "common/stuff.yml"
.funcs("common/stuff.yml") # returns module that has exported methods (see module package)
```

> Why does `yaml` takes an array and `func` does not?

we can append yaml documents together without losing anything. however, with functions we cannot "append" without losing functions that are named the same. i thought it was a safest thing for now, and we can iterate what does it mean to load a package vs a module.

`.text(["common/stuff.yml"])` is also a bit weird as appending plain text isnt that useful.

> Why not take varargs in the function (less syntax cluttering, but not important)?

being a bit more explicit and leaving possibility of adding future options as the last arg.

> Also, a feature I wanted in this module was the ability to get all the text or all the yaml from all files from the library if no argument was provided. it frees the library user from knowing the library internals.

i'd like that as well. just havent formed an opinion on what that should look like. goes back to above point about merging funcs with same names together. possibly extend those functions to support accepting just the package name (eg `.funcs(["common"])`) or no arguments at all to indicate everything.

probably should also rename `funcs` to `defs` to align it better with `def` identifier.

**Comment by mildred** [comment]
Well, the main question I had was about _ytt_lib and if nesting it would make sense... If a single _ytt_lib location should be the target for all loads (like GOPATH) or if instead we should nest them (like node_modules)

@nimakaviani, you seem to have answered in favor of the node_modules approach, that is a library loaded as `@github.com/lib1` that itself is loading a library `@github.com/lib2` would search for it in `/_ytt_lib/github.com/lib1/_ytt_lib/github.com/lib2`. Advantages are that you always encapsulate the dependency, but the drawback is that you end up with so many of them, with no unity.

The other approach would be to always load libraries in top level `_ytt_lib` folder only. The example above would look for the library in `/_ytt_lib/github.com/lib2` instead. @cppforlife, do you confirm that is not what you want ?

And this makes me think that perhaps separate from ytt itself, we could have a package manager which would populate _ytt_lib automatically, but that's something else entirely.

>> Also, a feature I wanted in this module was the ability to get all the text or all the yaml from all files from the library if no argument was provided. it frees the library user from knowing the library internals.
>
> I'd like that as well. just havent formed an opinion on what that should look like. goes back to above point about merging funcs with same names together. possibly extend those functions to support accepting just the package name (eg .funcs(["common"])) or no arguments at all to indicate everything.

Actually, getting the complete YAML output from a library is the main use case I was thinking of. Just as by default, when you execute ytt CLI on a directory it outputs everything and you are not forced to select which files to render

The idea of selecting a package name to get the output from it seems a nice idea, including the root package (indicated as "/" ?). Could it be the default if no option is provided or do you prefer it to be explicit (`.yaml()` would be a shortcut for `.yaml(["/"])`)

**Comment by nimakaviani** [comment]
> @nimakaviani, you seem to have answered in favor of the node_modules approach, that is a library loaded as @github.com/lib1 that itself is loading a library @github.com/lib2 would search for it in /_ytt_lib/github.com/lib1/_ytt_lib/github.com/lib2. Advantages are that you always encapsulate the dependency, but the drawback is that you end up with so many of them, with no unity.

yes I think i have been advocating for a self-contained modularized model from the get-go. The idea of the library being self-contained is very appealing to me. 

yes there is definitely the problem with each library potentially inheriting a copy of a dependency but again if they are all pulled from a unified source (like git for example) the resolution should be relatively easy. Also broken links and dealing with versioning of loaded modules becomes much less of a problem.

**Comment by cppforlife** [comment]
> Well, the main question I had was about _ytt_lib and if nesting it would make sense... If a single _ytt_lib location should be the target for all loads (like GOPATH) or if instead we should nest them (like node_modules)

my previous comment didn't explicitly mention it but yes, nesting of _ytt_lib is absolutely supported and encouraged to lock down correct dependencies. there is no need to searching for paths in this case either.

> Actually, getting the complete YAML output from a library is the main use case I was thinking of. Just as by default, when you execute ytt CLI on a directory it outputs everything and you are not forced to select which files to render

yup yup. just wanted to limit scope at the beginning.

> Could it be the default if no option is provided or do you prefer it to be explicit (.yaml() would be a shortcut for .yaml(["/"]))

i dont mind making it a default if no args are provided.

**Comment by pavel-khritonenko** [comment]
Not sure if it related to that particular PR, but it makes my situation solved. What I want - is to have some "templates" or "main" set component, would place that into root folder and then reuse in each particular implementation of that component, something like:

```components.lib.yaml
app1/deployment.yaml
app2/deployment.yaml
```

And in each of those deployments, I would refer to common component file, like:

```load("../components.lib.yaml", "deployment")```

However, it seems not possible yet (tested with `ytt` 0.18.0)

**Comment by cppforlife** [comment]
> And in each of those deployments, I would refer to common component file, like: load("../components.lib.yaml", "deployment")

correct it's not possible to use `../` in load paths.

thanks as well for letting us know that this will be helpful for your use case.

to unblock you for now, you can do the following (as prev proposed by @mildred in a different issue):

```
$ ls -l
app1
app2
shared.lib.yml

$ ls -l app1/
config.yml
shared.lib.yml -> ../shared.lib.yml

$ ls -l app2/
config.yml
shared.lib.yml -> ../shared.lib.yml
```

given above fs structure, you can use `--allow-symlink-destination` to allow "inside project" symlinks:

```
$ ytt -f . --allow-symlink-destination .
```

**Comment by jdiasamaro** [comment]
+1 nice feature to add. Very helpful indeed

**Comment by cppforlife** [comment]
https://github.com/k14s/ytt/releases/tag/v0.23.0 includes @ytt:library feature that yall may find appropriate for your use case /cc @carlosflorencio @tfpereira @jdiasamaro @ioben. since this issue covered a lot of use cases, let me know in slack or here if it fits your needs.

**Comment by cppforlife** [comment]
https://github.com/k14s/ytt/releases/tag/v0.24.0 includes ability to use / in load statement to load files relative to library root.

**Comment by pivotaljohn** [comment]
Noting:
- it's been more than a year since any progress of this issue (and related PRs are long closed/abandoned)
- several useful bits have been harvested, as noted above.

This "discussion" has become inactive (and is most decidedly _not_ in-progress). Closing.
